server/commitlog/commitlog.go

```diff
@@ -328,6 +328,55 @@ func (l *commitLog) OffsetForTimestamp(timestamp int64) (int64, error) {
 	return l.segments[len(l.segments)-1].NextOffset(), nil
 }
 
+// LatestOffsetBeforeTimestamp returns the latest offset whose timestamp is less
+// than or equal to the given timestamp.
+func (l *commitLog) LatestOffsetBeforeTimestamp(timestamp int64) (int64, error) {
```

Comment:
There's probably a much more efficient/concise way of implementing this, but I couldn't see it tonight

Reply:
Think I've tidied this up somewhat